Remove HMP from optimum-habana

jwieczorekhabana opened the pull request:
HMP is deprecated in favor of PyTorch autocast
- removed hmp usage
- removed setting autocast env variables through GaudiConfig
- updated tests
- updated docs
- updated README.md files

regisss commented:
@jwieczorekhabana Thanks for opening this PR! I'll review it in the beginning of next week and I'll try to accelerate on merging the autocast PRs.

regisss commented on optimum/habana/transformers/gaudi_configuration.py:
```diff
@@ -52,40 +52,13 @@ class GaudiConfig(BaseConfig):
-            )
-        self.hmp_opt_level = kwargs.pop("hmp_opt_level", "O1")
-
-    def write_bf16_fp32_ops_to_text_files(
-        self,
-        path_to_bf16_file: Path,
-        path_to_fp32_file: Path,
-    ):
-        for path, ops in zip(
-            [Path(path_to_bf16_file), Path(path_to_fp32_file)], [self.hmp_bf16_ops, self.hmp_fp32_ops]
-        ):
-            with path.open("w") as text_file:
-                # writelines does not add new lines after each element so "\n" is inserted
-                text_file.writelines(op + "\n" for op in ops)
```
We still need this method to be able to specify autocast custom op lists in a Gaudi config

regisss commented on tests/test_gaudi_configuration.py:
```diff
@@ -50,36 +26,6 @@ class GaudiConfigTester(unittest.TestCase):
-        self.assertTrue(is_list_of_strings(gaudi_config.hmp_bf16_ops))
-        self.assertTrue(is_list_of_strings(gaudi_config.hmp_fp32_ops))
-
-    def test_write_bf16_fp32_ops_to_text_files(self):
-        gaudi_config = GaudiConfig()
-
-        with tempfile.NamedTemporaryFile() as bf16_file:
-            with tempfile.NamedTemporaryFile() as fp32_file:
-                gaudi_config.write_bf16_fp32_ops_to_text_files(
-                    bf16_file.name,
-                    fp32_file.name,
-                )
-
-                self.assertTrue(
-                    filecmp.cmp(
-                        bf16_file.name,
-                        BF16_OPS_REFERENCE_FILE,
-                        shallow=False,
-                    )
-                )
-                self.assertTrue(
-                    filecmp.cmp(
-                        fp32_file.name,
-                        FP32_OPS_REFERENCE_FILE,
-                        shallow=False,
-                    )
-                )
```
We should keep this test as `write_bf16_fp32_ops_to_text_files` will still be used to specify custom op lists.
I propose to instantiate a Gaudi config such as:
```py
gaudi_config = GaudiConfig(
    autocast_bf16_ops=[
        "add",
        "addmm",
        "bmm",
        "div",
        "dropout",
        "gelu",
        "iadd",
        "linear",
        "layer_norm",
        "matmul",
        "mm",
        "rsub",
        "softmax",
        "truediv",
    ],
    autocast_fp32_ops=[
        "embedding",
        "nll_loss",
        "log_softmax",
    ],
)
```

regisss commented on docs/source/package_reference/gaudi_config.mdx:
```diff
@@ -16,69 +16,17 @@ limitations under the License.
-- `hmp_is_verbose` enables to decide whether to log precision decisions for each operation for debugging purposes. It is disabled by default. You can find an example of such log [here](https://docs.habana.ai/en/latest/PyTorch/PyTorch_Mixed_Precision/PT_Mixed_Precision.html#hmp-logs).
 - `use_fused_adam` enables to decide whether to use the [custom fused implementation of the ADAM optimizer provided by Habana](https://docs.habana.ai/en/latest/PyTorch/Model_Optimization_PyTorch/Custom_Ops_PyTorch.html#custom-optimizers).
 - `use_fused_clip_norm` enables to decide whether to use the [custom fused implementation of gradient norm clipping provided by Habana](https://docs.habana.ai/en/latest/PyTorch/Model_Optimization_PyTorch/Custom_Ops_PyTorch.html#other-custom-ops).
-- `hmp_bf16_ops` enables to specify the Torch operations that should be computed in *bf16*. You can find more information about casting rules [here](https://docs.habana.ai/en/latest/PyTorch/PyTorch_Mixed_Precision/PT_Mixed_Precision.html#basic-design-rules).
-- `hmp_fp32_ops` enables to specify the Torch operations that should be computed in *fp32*. You can find more information about casting rules [here](https://docs.habana.ai/en/latest/PyTorch/PyTorch_Mixed_Precision/PT_Mixed_Precision.html#basic-design-rules).
```
I think we should mention:
- `use_torch_autocast` but saying that `--bf16` should be favored as `use_torch_autocast` is used to define a good pre-defined config
- `autocast_bf16_ops` and `autocast_fp32_ops` as #308 enables users to specify cutom op lists but saying that the default should work for most models

regisss replied:
As discussed by email, regarding `autocast_bf16_ops` and `autocast_fp32_ops`, I'm fine with saying that the env variable way should be favored. But they should still be documented.

regisss commented on docs/source/usage_guides/accelerate_training.mdx:
```diff
@@ -57,44 +57,16 @@ To not take them into account in the computation of the throughput at the end of
 
 Mixed-precision training enables to compute some operations using lighter data types to accelerate training.
-Habana Mixed Precision (HMP) proposes to mix *fp32* and *bf16* operations.
+Optimum-Habana enables mixed precision training in a similar fasion as HuggingFace transofrmers.
```
```suggestion
Optimum Habana enables mixed precision training in a similar fashion as 🤗 Transformers.
```

regisss commented on docs/source/usage_guides/accelerate_training.mdx:
```diff
@@ -57,44 +57,16 @@ To not take them into account in the computation of the throughput at the end of
+- argument `--bf16` enables usage of PyTorch autocast
+- argument `--half_precision_backend [hpu_amp, cpu_amp]` is used to specify a device on which mixed precision operations should be performed
 
-<Tip warning={true}>
-
-Please refer to the [list of supported PyTorch operators](https://docs.habana.ai/en/latest/PyTorch/Pytorch_Operators/Pytorch_Operators.html) beforehand to make sure the ones you are interested in are compatible with *bf16*.
 
-</Tip>
```
I would keep this

jwieczorekhabana replied:
But those operators are incompatible with autocast. HMP and autocast operate on different software levels. Please see: https://docs.habana.ai/en/latest/PyTorch/PyTorch_Mixed_Precision/Autocast.html#override-options

regisss replied:
No problem, we don't have to keep the same operators. Maybe it will just be easier to refer to GPT2's Gaudi config.

regisss commented on docs/source/usage_guides/accelerate_training.mdx:
````diff
@@ -57,44 +57,16 @@ To not take them into account in the computation of the throughput at the end of
-```
-
-<Tip>
+<Tip warning={true}>
````
```suggestion
<Tip>
```
I think this one doesn't have to be a warning, it's rather extra information

jwieczorekhabana replied:
I'd rather keep it as a warning as the names of operations in HMP and autocast may differ. Wouldn't it be important for users switching from HMP to autocast.

regisss commented on docs/source/usage_guides/accelerate_training.mdx:
````diff
@@ -57,44 +57,16 @@ To not take them into account in the computation of the throughput at the end of
-</Tip>
-
-To apply HMP, you must set `"use_habana_mixed_precision"` to `true` in the Gaudi configuration file.
-Then, you can specify which operators to compute in *bf16* with `"hmp_bf16_ops"` and which operators to compute in *fp32* with `"hmp_fp32_ops"`.
-If these operators are not specified, their default values are set to be the ones written in the [Gaudi configuration file of BERT](https://huggingface.co/Habana/bert-large-uncased-whole-word-masking/blob/main/gaudi_config.json), which is a good starting point for applying HMP:
-```
-"hmp_bf16_ops": [
-    "add",
-    "addmm",
-    "bmm",
-    "div",
-    "dropout",
-    "gelu",
-    "iadd",
-    "linear",
-    "layer_norm",
-    "matmul",
-    "mm",
-    "rsub",
-    "softmax",
-    "truediv"
-],
-"hmp_fp32_ops": [
-    "embedding",
-    "nll_loss",
-    "log_softmax"
-]
-```
````
I would still keep a part of this to show how to specify custom op lists. We can add a link to the GPT2 Gaudi config when it is updated.

jwieczorekhabana replied:
But shouldn't users provide custom lists in a similar way to other training demos outside of HuggingFace? We can keep those in GaudiConfig to make sure they are optimized for specific model.

regisss reviewed (commented):
@jwieczorekhabana I left several comments, the main point being that #308 enables users to specify their own op lists in Gaudi config files so this should be documented, tested, etc (which should basically consist in changing a few variable names).

For OPT, I need to check if we need to disable autocast for this specific line.

**Note that we should not merge this PR before we manage to make Wav2Vec2 work with autocast.** It is being investigated but no fix has been merged yet.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-habana/pr_349). All of your documentation changes will be reflected on that endpoint.

regisss commented on docs/source/package_reference/gaudi_config.mdx:
```diff
@@ -16,69 +16,19 @@ limitations under the License.
-
-</Tip>
+- `use_torch_autocast` enables PyTorch autocast; used to define good pre-defined config; users should favor --bf16 training argument
+- `autocast_bf16ops` list of operations that should be run with bf16 precision under autocast context
+- `autocast_fp32_ops` list of operations that should be run with fp32 precision under autocast context
```
You can specify here that setting the env variables `LOWER_LIST` and `FP32_LIST` is the prioritized way of specifying these ops

regisss commented on docs/source/usage_guides/accelerate_training.mdx:
```diff
@@ -57,44 +57,16 @@ To not take them into account in the computation of the throughput at the end of
 
 Mixed-precision training enables to compute some operations using lighter data types to accelerate training.
-Habana Mixed Precision (HMP) proposes to mix *fp32* and *bf16* operations.
+Optimum Habana enables mixed precision training in a similar fasion as 🤗 Transofrmers.
```
```suggestion
Optimum Habana enables mixed precision training in a similar fasion as 🤗 Transofrmers:
```

jwieczorekhabana replied:
done

regisss commented on tests/test_gaudi_configuration.py:
```diff
@@ -13,35 +13,11 @@
 from optimum.habana import GaudiConfig
 
 
-BF16_OPS_REFERENCE_FILE = Path(__file__).parent.resolve() / Path("configs/bf16_ops.txt")
-FP32_OPS_REFERENCE_FILE = Path(__file__).parent.resolve() / Path("configs/fp32_ops.txt")
```
We still need those in `test_write_bf16_fp32_ops_to_text_files`

jwieczorekhabana replied:
done

regisss commented on optimum/habana/transformers/models/gpt2/modeling_gpt2.py:
```diff
@@ -331,7 +331,6 @@ def gaudi_gpt2_forward(
     """
     Copied from GPT2Model.forward: https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py
     The only differences are:
-    - disable HMP cast for attention_mask
```
Let's keep this line and change "HMP" by "autocast"

regisss commented on optimum/habana/diffusers/pipelines/pipeline_utils.py:
```diff
@@ -109,16 +109,13 @@ def __init__(
                     )
                     self.gaudi_config.use_torch_autocast = False
-                    self.gaudi_config.use_habana_mixed_precision = False
-                elif self.gaudi_config.use_torch_autocast:
-                    # Open temporary files to write mixed-precision ops
```
We need to keep these two lines for autocast as `bf16_full_eval` doesn't rely on autocast (it basically converts everything into bf16)

jwieczorekhabana replied:
done

regisss reviewed (commented):
Thanks for updating the PR! I left a few comments.

Could you also run the following to make the code style check pass please?
```
pip install black ruff
make style
```

regisss commented on docs/source/usage_guides/accelerate_training.mdx:
```diff
@@ -57,7 +57,7 @@ To not take them into account in the computation of the throughput at the end of
 
 Mixed-precision training enables to compute some operations using lighter data types to accelerate training.
-Optimum Habana enables mixed precision training in a similar fasion as 🤗 Transofrmers.
+Optimum Habana enables mixed precision training in a similar fasion as 🤗 Transofrmers:
```
```suggestion
Optimum Habana enables mixed precision training in a similar fashion as 🤗 Transformers:
```

regisss reviewed (approved):
LGTM!

Let's wait to know what happens with Wav2Vec2 training before merging anything cc @libinta

regisss commented:
@jwieczorekhabana will the Gaudi config for Wav2Vec2 be the following?
```json
{
  "use_fused_adam": true,
  "use_fused_clip_norm": true,
  "use_torch_autocast": true
}
```
Or will there be any specific bf16 ops?

jwieczorekhabana commented:
> @jwieczorekhabana will the Gaudi config for Wav2Vec2 be the following?
> 
> ```json
> {
>   "use_fused_adam": true,
>   "use_fused_clip_norm": true,
>   "use_torch_autocast": true
> }
> ```
> 
> Or will there be any specific bf16 ops?

That'll be it. There is a hanging PR on HuggingFace https://huggingface.co/Habana/wav2vec2/discussions/2/files that can be merged after 1.13 release.
Similarly for gpt2 there is a closed PR that can be reopened https://huggingface.co/Habana/gpt2/discussions/4/files

Release 1.13 addresses issues both of those topologies had.

regisss commented:
Cool, I'll merge them when 1.13 is released then :+1:

regisss commented:
@jwieczorekhabana Can you rebase this branch on main? There are merge conflicts in `tests/test_diffusers.py` so GitHub won't let me merge the PR when 1.13 is released.

jwieczorekhabana commented:
> @jwieczorekhabana Can you rebase this branch on main? There are merge conflicts in `tests/test_diffusers.py` so GitHub won't let me merge the PR when 1.13 is released.

@regisss Done

regisss commented:
@jwieczorekhabana It seems there are still some merge conflicts in `tests/test_diffusers.py`

regisss commented:
@jwieczorekhabana I pushed a few commits to solve the CI that was not passing
